projects/ad405x_iio/app/ad405x_iio.c

```diff
@@ -0,0 +1,1576 @@
+static struct iio_attribute iio_ad405x_global_attributes[] = {
+	AD405X_CHN_ATTR("operating_mode", ADC_OPERATING_MODE),
+	AD405X_CHN_AVAIL_ATTR("operating_mode_available", ADC_OPERATING_MODE),
+#if (ADC_CAPTURE_MODE != SAMPLE_MODE)
+#if (ADC_CAPTURE_MODE == BURST_AVERAGING_MODE)
+	AD405X_CHN_ATTR("burst_sample_rate", ADC_BURST_SAMPLE_RATE),
+	AD405X_CHN_AVAIL_ATTR("burst_sample_rate_available", ADC_BURST_SAMPLE_RATE),
+#endif
+	AD405X_CHN_ATTR("avg_filter_length", ADC_FILTER_LENGTH),
+	AD405X_CHN_AVAIL_ATTR("avg_filter_length_available", ADC_FILTER_LENGTH),
+#endif
+	AD405X_CHN_ATTR("sampling_frequency", ADC_SAMPLE_RATE),
+	END_ATTRIBUTES_ARRAY,
+};
+
+/* IIO channels info */
+static struct iio_channel iio_ad405x_channels[] = {
+	{
+#if (ADC_CAPTURE_MODE == SAMPLE_MODE)
+		.name = "sample_mode",
+#elif (ADC_CAPTURE_MODE == BURST_AVERAGING_MODE)
+		.name = "burst_average_mode",
+#elif (ADC_CAPTURE_MODE == AVERAGING_MODE)
+		.name = "averaging_mode",
+#endif
+		.ch_type = IIO_VOLTAGE,
+		.ch_out = false,
+		.indexed = true,
+		.channel = 0,
+		.scan_index = 0,
+		.scan_type = &ad405x_iio_scan_type,
+		.attributes = iio_ad405x_ch_attributes
+	}
+};
+
+/* ad405x IIOD debug attributes list */
+static struct iio_attribute ad405x_debug_attributes[] = {
+	END_ATTRIBUTES_ARRAY
+};
+
+/* Global Pointer for IIO Device Data */
+volatile struct iio_device_data* iio_dev_data_g;
+
+/* Global variable for number of samples */
+uint32_t nb_of_samples_g;
+
+/* Global variable for data read from CB functions */
+int32_t data_read;
+
+/* SPI Message */
+struct no_os_spi_msg ad405x_spi_msg;
+
+/* SPI init params */
+struct stm32_spi_init_param* spi_init_param;
+
+/******************************************************************************/
+/************************** Functions Declarations ****************************/
+/******************************************************************************/
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+/*!
+ * @brief	Function to configure pwm period
+ * @param	sampling_pwm_period[in] - Time period of sampling PWM.
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int configure_pwm_period(uint32_t sampling_pwm_period)
+{
+	int ret;
+#if (INTERFACE_MODE == SPI_DMA)
+	cs_init_params.period_ns = CONV_TRIGGER_PERIOD_NSEC(ad405x_sample_rate);
+	pwm_init_params.period_ns = CONV_TRIGGER_PERIOD_NSEC(ad405x_sample_rate);
+	ret = init_pwm();
+	if (ret) {
+		return ret;
+	}
+#else
+#if (ACTIVE_PLATFORM == MBED_PLATFORM)
+	ret = no_os_pwm_enable(pwm_desc);
+	if (ret) {
+		return ret;
+	}
+#endif
+
+	ret = no_os_pwm_set_period(pwm_desc,
+				   sampling_pwm_period);
+	if (ret) {
+		return ret;
+	}
+
+	/* Note: Duty Cycle in SPI DMA mode is fixed for any sampling rate */
```

Comment:
Is this not SPI Interrupt condition here?

Reply:
it is indeed. But with this comment I just wanted to make clear to the reader that duty cycle is being changed only in case of the interrupt mode. And that it's not an oversight, but very much intentional